proposals/p6333.md

```diff
@@ -0,0 +1,586 @@
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/6333)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+    -   [Look-and-feel](#look-and-feel)
+-   [Details](#details)
+    -   [Command changes](#command-changes)
+        -   [Compile command](#compile-command)
+        -   [Build command](#build-command)
+        -   [Link command](#link-command)
+    -   [carbon_library and carbon_binary](#carbon_library-and-carbon_binary)
+        -   [Indirect API exposure](#indirect-api-exposure)
+        -   [Core package rules](#core-package-rules)
+    -   [Mapping packaging directives to filenames](#mapping-packaging-directives-to-filenames)
+        -   [Disallow ambiguous library names](#disallow-ambiguous-library-names)
+-   [Future work](#future-work)
+    -   [Caching checked IR, C++ AST, and other possible compile artifacts](#caching-checked-ir-c-ast-and-other-possible-compile-artifacts)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Naming of commands and rules](#naming-of-commands-and-rules)
+    -   [apis versus hdrs](#apis-versus-hdrs)
+    -   [Support a full-fledged build system](#support-a-full-fledged-build-system)
+    -   [Don't support packaging directive to filename mappings](#dont-support-packaging-directive-to-filename-mappings)
+    -   [Distribute pre-compiled versions of Core files](#distribute-pre-compiled-versions-of-core-files)
+    -   [Require C++ files be passed to clang directly](#require-c-files-be-passed-to-clang-directly)
+    -   [Allow indirect dependencies in carbon_library](#allow-indirect-dependencies-in-carbon_library)
+    -   [Implicit carbon_library dependency on Core versus prelude](#implicit-carbon_library-dependency-on-core-versus-prelude)
+    -   [Create an explicit mapping from packaging directives to files](#create-an-explicit-mapping-from-packaging-directives-to-files)
+
+<!-- tocstop -->
+
+## Abstract
+
+-   Change the look-and-feel of the `carbon` compilation command set to use
+    `compile`, `link`, and `build`.
+-   Rewrite `carbon_library` and `carbon_binary` based on the `compile` and
+    `link` commands.
+-   Build library-to-file discovery for `Core`, but support it in a general
+    manner.
+
+## Problem
+
+Right now, `carbon compile` produces an object file per input file. This has a
+few issues, for example when it comes to compiling prelude files: it's not clear
+how to get a prelude `.o` file.
+
+Additionally, the compilation setup doesn't work well with the prelude because
+symbols don't get emitted for prelude files. Right now, we generally don't
+notice this because the prelude has been implemented in a way that typically
+doesn't require separate object files, but it will likely become an issue.
```

Comment:
For what it's worth, this *does* work these days with the existing `carbon_library` rule, because it adds all the prelude files to the list of sources. Some of our examples depend on that. But as you mention in the next paragraph, that doesn't really apply to real-world build scenarios where it won't be easy to get a list of the prelude files and build them manually.